onnxruntime/core/providers/nnapi/nnapi_builtin/builders/model_builder.cc

```diff
@@ -186,6 +186,33 @@ void ModelBuilder::PreprocessInitializers() {
   }
 }
 
+// Help to get all quantized operators' input and the node(s) using the input
+std::unordered_map<std::string, vector<const Node*>> GetAllQuantizedOpInputs(const GraphViewer& graph_view) {
+  std::unordered_map<std::string, vector<const Node*>> all_quantized_op_inputs;
+  const auto& node_indices = graph_view.GetNodesInTopologicalOrder();
+  for (const auto& node_idx : node_indices) {
+    const auto* node(graph_view.GetNode(node_idx));
+    const auto& op_type = node->OpType();
+    if (op_type == "DequantizeLinear" || op_type == "QLinearMatMul" || op_type == "QLinearConv") {
+      const auto& input_name = node->InputDefs()[0]->Name();
+      if (Contains(all_quantized_op_inputs, input_name))
+        all_quantized_op_inputs.at(input_name).push_back(node);
+      else
+        all_quantized_op_inputs.emplace(input_name, vector<const Node*>{node});
+    }
+
+    if (op_type == "QLinearMatMul" || op_type == "QLinearConv") {
+      const auto& input_name = node->InputDefs()[3]->Name();
```

Comment:
Would be nice to just do the string comparison on op_type once. Could put the input name processing in a lambda which would help with that.

Reply:
I will add a helper function to convert op_type to an enum to handle this in next PR, will keep the string comparison for now

Reply:
I will add a helper function to convert op_type to an enum to handle this in next PR, will keep the string comparison for now



---
In reply to: [456962221](https://github.com/microsoft/onnxruntime/pull/4530#discussion_r456962221) [](ancestors = 456962221)